presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -241,16 +338,23 @@ private JoinNode chooseJoinOrder(JoinGraphNode joinGraph)
             if (join == null) {
                 join = generatePartitions(joinGraph.getSources().size())
                         .map(partitioning -> createJoinNodeTree(joinGraph, partitioning, idAllocator))
+                        .filter(Optional::isPresent)
+                        .map(Optional::get)
```

Comment:
Now this will throw `IllegalStateException` if there isn't a join enumeration without a cross join. You should relax condition at: https://github.com/Teradata/presto/pull/559/commits/3db11ff4e1f4da66913ae65dc33e1b49b2940ca0#diff-5cdb22a9680530f837d520839d7eaf99L245

Additionally, you should memoize information that for given `JoinGrapNode` there isn't any order without cross join so that you don't check for cross joins multiple times for the same source nodes.